### DISCUSSION TITLE
Feedback on using osxphotos to backup a large photo library to a NAS

### DISCUSSION THREAD

**Discussion by gerdemb** [RhetTbull/osxphotos#1722] [General]
Hello!

I've been testing osxphotos to back up my photo library to my NAS, and I wanted to share some feedback and my experiences with the tool. My use case is to back up my entire library to my NAS. I have 267,296 items in my library that take up 1.3TB of storage in iCloud with "Optimize Storage" enabled.

First off, I've been looking for a tool like this for a long time and I want to acknowledge the incredible effort put in by the community to make osxphotos as powerful and flexible as it is.  I want to emphasize that my comments are intended as constructive criticism, to help make osxphotos even better.

Initially, I found the number of command-line options a bit overwhelming as a beginner. I ended up reading all the documentation before I felt confident I had the correct combination of switches to do a full backup; however, I still wasn't entirely sure I had the right combination and it seems like the that combination could be a moving target as new versions are released. I think osxphotos would benefit from an option similar to rsync's `--archive`, which provides a sensible default combination of switches. The use case of "export my entire library while preserving as much information as possible" seems pretty standard. For reference, here's the combination of switches I settled on:

```
export /Volume/NAS/photos --export-by-date --sidecar xmp --update --export-aae --person-keyword --download-missing --use-photo-kit
```

Related GitHub discussions:
[https://github.com/RhetTbull/osxphotos/discussions/1267](https://github.com/RhetTbull/osxphotos/discussions/1267)
[https://github.com/RhetTbull/osxphotos/discussions/1659](https://github.com/RhetTbull/osxphotos/discussions/1659)
[https://github.com/RhetTbull/osxphotos/discussions/1515](https://github.com/RhetTbull/osxphotos/discussions/1515)

Another issue I encountered was with performance and progress reporting. I ran this command continuously for several hours every day for several weeks, but it never seemed to come close to completing a full export. At the beginning of each run, the export always started from zero, making it unclear how much had been completed in previous runs. On each subsequent run, I could see more "skipped up-to-date file" messages, but it seemed to take just as much time to "skip" a file as it did to back up a new one. In the end, the highest number of completed files exported was around 20,000, or about 10% of my library, despite running for 6-8 hours per day for a couple weeks. Eventually, I gave up, as it was never able to complete a full backup. I realize my library is probably  larger than average, which could have contributed to the performance issues. Or perhaps I had some setting wrong?

Finally, (and I realize this is a small nitpick), CTRL-C wouldn't stop the export on my machine, and I always had to force-kill the process.

Thanks again to the developers and community for all the effort that has gone into making osxphotos such a useful tool. I hope my feedback can help improve it further!

Cheers,
Ben

As an aside, after failing to complete a full backup with osxphotos, I created a simple photo backup app inspired by Time Machine. It uses the Apple PhotoKit API, so it lacks the advanced features that osxphotos has by reverse-engineering the Photos database, but it offers much faster performance for my large library. If you're curious, there's a link in my GitHub profile.

**Comment by RhetTbull** [comment]
Thanks for the feedback. On a NASA you must use `--ramdb` and or specify a different export database path using `--exportdb`.

OSXPhotos is not optimized for NAS. There is an open issue to have NAS auto detected but this is low priority. OSXPhotos is also not optimized for Optimize Mac Storage. 

You could probably speed this up with `--ignore-signature` 

I'm sorry it never completed for you. I've successfully exported larger libraries than yours with no problem.  But I only use fast locally and attached storage.

**Comment by gerdemb** [comment]
Thanks for responding. I added the flags `--ramdb` `--exportdb /local/dir` `--ignore-signature` and `--checkpoint 100` and am running the export again. I'll report back in a day or two the results. The `osxphotos_export.db` on my NAS was 1.23GB.

**Reply by RhetTbull** [reply]
Using --checkpoint 100 means the 1.2GB file has to be written to disk every 100 images. You're probably better off with a bigger number.

**Comment by oPromessa** [comment]
@gerdemb 

I've been using osxphotos on a 100K pics Library into NAS Synology successfully for a few year already 😄 !

I also use --cleanup to delete any pics on the exported folder that I happen to delete from Photos.

I don't use --photo-kit.

I don't use iCloud storage or the need for downloading images from iCloud due to the "Optimize Mac Storage"... which I believe can slow things down.

First backup can take time.. especially if you have videos. 

I have 1 GBPs network and my exports can take on first export,or when a large number of files need to be updated, 8~10 hours for ~100K pics. but subsequent daily's with a few thousands pics changed requiring update/copy sits around 4 hours.

From these somehow outdated tests (https://github.com/RhetTbull/osxphotos/issues/582#issuecomment-1053623946) on 1GB LAN it was taking about 0.3 seconds per pic (not huge Videos!) which would yield ~16.6 hours for 200K pics.

If you incorporate --verbose and --timestamp options you get an output like the following and track progress and performance. 

In the example below it tells you:
* for example exiftool plus copy of file took from second 11.212 to 11.488 so about 0.276 seconds.
* progress is indicted with pic 89000 of 107186

```py
(...)
2024-10-21 16:27:25.604177 -- Exporting XXX.HEIC (XXX.HEIC) (89000/107186)
2024-10-21 16:27:25.697627 -- Skipped up to date file /Volumes/photo/XXX.HEIC
2024-10-21 16:27:25.705689 -- Exporting ZZZ.JPG (ZZZ.JPG) (89001/107186)
2024-10-21 16:27:25.807583 -- Skipped up to date file /Volumes/photo/ZZZJPG
(...)
2024-10-21 16:27:11.212614 -- Exporting IMG_WRITTEN.jpg (IMG_WRITTEN.jpg) (107082/107186)
2024-10-21 16:27:11.327217 -- Writing metadata with exiftool for IMG_WRITTEN.jpg
2024-10-21 16:27:11.488014 -- Exported IMG_WRITTEN.jpg.jpg to
(...)
2024-10-21 16:27:26.256606 -- Elapsed time: 4:36:09
(...)
2024-10-21 16:27:26.303988 -- Cleaning up /Volumes/photo
2024-10-21 16:31:36.375673 -- Deleting /Volumes/photo/TTTjpeg
2024-10-21 16:32:59.701735 -- Deleted: 1 file, 0 directories
```

**Reply by gerdemb** [reply]
@oPromessa Thanks for the tip on the `--timestamp` option! I'm going to add it to my export command.

To be clear, I fully expected the first complete backup to take many days to complete, but with every export run starting back at zero and never getting past more than ~20k pictures in a day, it seemed like it was never going to finish.

I understand that "Optimize Storage" requires downloads from iCloud to get missing data (which I imagine significantly slows down the export), but I can't fit my entire library on my MacBook. However, if you can fit your entire photo library on a local drive why not just backup the Photos Library directly instead of using OSXPhotos?

**Reply by oPromessa** [reply]
Further to this I have a small python script which detect lengthy operations (over 60secs for instance) from a --verbose/--timestamp osxphotos log file:

I run `caffeinate -s osxphotos` to prevent the Mac from sleeping but I have noticed that even with caffeinate, on occasion, the Mac slows down the activity... Mystery to me! 

For my case, you can see in some cases a exiftool+copy of 214 seconds... this was for a 1.3 GB MP4 videos

```py
Sequence:   1 Line:    234261 over [ 60 secs]      75.45 Duration: 00:01:15.449956 2024-10-20 04:15:01.670564 -- Writing metadata with exiftool for XXX.MOV
Sequence:   1 Line:    252845 over [ 60 secs]      68.53 Duration: 00:01:08.525544 2024-10-20 04:49:53.517625 -- Writing metadata with exiftool for XXX.MP4
Sequence:   1 Line:    274033 over [ 60 secs]     214.69 Duration: 00:03:34.686064 2024-10-20 05:29:36.144517 -- Writing metadata with exiftool for XXX.MP4
Sequence:   1 Line:    282734 over [ 60 secs]      69.05 Duration: 00:01:09.050918 2024-10-20 05:47:40.673230 -- Writing metadata with exiftool for XXX.mp4
```

If one compares with  rsync for the same file and origin/destination.. it's much faster. But then again it is not performing several databases and comparisons operations required and also applying/using exiftool changes!

```
rsync  --progress  /Users/Shared/Pictures/Pictures/iPhoto/XXX.MP4 /Volumes/photo/XXX.MP4
  1438764361 100%   58.06MB/s    0:00:23 (xfer#1, to-check=0/1)

sent 1438940079 bytes  received 42 bytes  58732249.84 bytes/sec
total size is 1438764361  speedup is 1.00
```

**Reply by oPromessa** [reply]
> To be clear, I fully expected the first complete backup to take many days to complete, but with every export run starting back at zero and never getting past more than ~20k pictures in a day, it seemed like it was never going to finish.

Strange.. it should not start from zero!

You can try the --limit option to incrementally backup. osxphotos would then pick/up from where it left-off -- I still think it must check no changes occurred on previously exported files but if would be faster.
```
--limit LIMIT                   Export at most LIMIT photos. Useful for
                                  testing. May be used with --update to export
                                  incrementally.
```

@RhetTbull doesn't like it -- joking 😄 -- but I also use --exportdb and keep the database local.  Avoids a few seconds locking it up into memory and writing it back to disk at the end. But during the copy should not make a difference.

Look at --retry option to accommodate for transient copy over the LAN errors.

Also note the --cleanup option should you choose to use it deletes *all* unknown files.  Use the --keep if you have any files you want to keep on the export folder which are not Pics/Movies related.
 

> I understand that "Optimize Storage" requires downloads from iCloud to get missing data (which I imagine significantly slows down the export), but I can't fit my entire library on my MacBook. However, if you can fit your entire photo library on a local drive why not just backup the Photos Library directly instead of using OSXPhotos?

In my case I export into [Synology Photos](https://www.synology.com/en-global/dsm/feature/photos) and is not only for backup purposes. With Synology Photos I can make all the pics replicating the Photos structure and nicely organised in Albums, tagged with keywords (except videos), geolocation, etc.. and make it available via the Web (behind CloudFlare) and via Mobile app Synology Photos Mobile for the whole family to access on-the-go anytime anywhere.

**Reply by RhetTbull** [reply]
> However, if you can fit your entire photo library on a local drive why not just backup the Photos Library directly instead of using OSXPhotos?

You're missing the point of OSXPhotos. It's designed to export photos and metadata so you have an archive copy *outside* the Apple ecosystem. Backing up the Photos library still keeps your photos locked in a proprietary database. 

OSXPhotos also gives you full control over the output. 

If what you really want to do is just backup from iCloud then [iCloud Photos Downloader](https://github.com/icloud-photos-downloader/icloud_photos_downloader) might be a better option.

**Comment by gerdemb** [comment]
@RhetTbull and @oPromessa thank you again for your suggestions and feedback. I let OSXPhotos run overnight with the following command line:

```
osxphotos \
  export \
  /Volumes/photo/osxphotos \
  --export-by-date \
  --sidecar xmp \
  --update \
  --export-aae \
  --person-keyword \
  --download-missing \
  --use-photokit \
  --ramdb \
  --exportdb /Users/gerdemb/osxphotos_export.db \
  --ignore-signature \
  --checkpoint 100 \
  --timestamp \
  --verbose --verbose
  ```
  
Which backed up 11987 files in 11 hours and 28 minutes which is about 17 seconds per file. I'm going to start a second export run this morning (changing to `--checkpoint 1000`) to see what happens. 

@oPromessa  
> Strange.. it should not start from zero!

When I said "start from zero" I mean that every export always starts with a line like:

```
2024-10-24 18:24:33.177947 -- Exporting IMG_7640 1.HEIC (53C664AF-B79F-4BF2-BD74-FE6EF51451DE.heic) (1/267296)
```

So it's not possible to see how much of the library has already been exported by previous runs. I thought that each subsequent run would be slightly faster as it should be skipping files that have already been exported--and I can see messages in the log that it is skipping files--but the speed of the export seems to be roughly the same even with skipped files meaning that it never reaches the end. 

@RhetTbull 

Thanks for the tip about iCloud Photos Downloader. I had already experimented with it before discovering OSXPhotos. I was uncomfortable with the stability of their approach which is basically screen-scrapping  iCloud.com and concerned about the potential for getting my account locked for using it. (Probably being paranoid, but it's certainly not an "Apple approved" method for downloading your photos). I think OSXPhotos has a more robust architecture and uses the official PhotoKIt API (or AppleScript) for downloading photos. Also, OSXPhotos is the only tool I'm aware of that has an import feature that allows importing an exported archive of photos back into a Photo library.

**Reply by oPromessa** [reply]
Yup 17 seconds per file whereas I have 0.3 seconds is quite a big difference.

This is a bit beyond the scope of osxphotos -- as Rhet mentioned `OSXPhotos is not optimized for NAS. `... but I thing something's fishy here 😄 !

Again, unless your files are quite big Videos... 

You may have to check networking... there are info out there on optimising LAN to NAS... and also smoother optimisations.

* Are you on 1 Gbit cable connected and not Wifi?
   * On your Mac: Settings-Network-Ethernet-Advanced-Hardware you should see Speed: 1000baseT and Duplex:full-duplex
   * On your NAS to you should also be able to check the speed which is configured.
* Try also a rsync of a few hundred MB's file to see your throughput. 
* Take a look at Mac Activity Monitor and see LAN Throughput
* NAS normally also have some kind of performance monitor whee you could see the current transfer rate.

**Reply by RhetTbull** [reply]
You could try using `--profile` and `--limit 5000` to run the profiler to see where the code is spending its time. My guess is it's reading directories off the NAS to see which files are present (this is done for each file to avoid file naming collisions). 

The `--profile` is a global option which means it comes before the command: `osxphotos --profile export ...`

**Reply by RhetTbull** [reply]
See also #1086

**Reply by gerdemb** [reply]
My second export run this morning averaged ~3 seconds per file, which was what I was expecting after spot-checking the log as it was running. Perhaps my previous overnight run got stuck on something or the laptop went to sleep? Anyway, I'm trying a third run with `--profile` and `--limit 5000` to see if I can get some more detailed information.

In the meantime, I think the conversation has veered off into the performance details whereas as my problem is more with the progress reporting and/or lack of forward progress on the export. I fully expected the first backup to take a long time and TBH performance is not critical and I don't really mind if it takes 10 days or 40 days to complete

The problem is that progress reporting doesn't tell me how much of the backup has been completed and it _seems_ like there is no forward progress. For example, I expected that exports were "resumable" and after the first partial export, subsequent exports would report something like "100,000 files already exported, exporting remaining 200,000 files". Instead subsequent exports start again from "zero" and skip over the already exported files, but skipping over the exported files doesn't seem any faster then exporting them in the first place, and it's difficult to tell if we are making any forward progress on the full backup.

My use case is the same as @torarnv in this comment https://github.com/RhetTbull/osxphotos/issues/1086#issuecomment-2106240353

From reading through that thread it seems like this is a use case the OSXPhotos doesn't support well?

**Reply by RhetTbull** [reply]
> For example, I expected that exports were "resumable" and after the first partial export, subsequent exports would report something like "100,000 files already exported, exporting remaining 200,000 files".

Exports are resumable. But OSXPhotos still needs to check each photo to see if something has changed between Photos and the exported files. For example, you edited a photo or added a keyword. My priority is maximum preservation of data, not speed. To do this OSXPhotos must look at the exported file to see if it's different than expected. Over a slow network/NAS this takes time. OSXPhotos is not optimized for this use case. 

You can use `--only-new` to only evaluate files that were not previously exported. That may better match your definition of resumable. 

There are inefficiencies in the design that are particularly bad for your use case (optimized library + NAS). The previously mentioned issue is to track this and I'll eventually get around to looking at it but it's not a priority as it doesn't affect my daily use of OSXPhotos. (And I only have a few hours per week to work on this).  I'm happy to consider PRs or even suggestions to improve the checks in the code. 

As for never finishing, I don't know what the issue is but if there is a bug I'd definitely like to squash it. Unfortunately without being able to replicate it it's hard to know what the issue is. If OSXPhotos crashes it creates a crash log with debug info but it sounds like in your case it's just hanging.

**Comment by cfc62** [comment]
There are some very very odd networking issues going on with Sequoia right now. I've experienced two, one of which I'll describe at the end for the curious. 

I've also seen many reports of users having issues with NASs and Sequoia. Googling "Sequoia Networking" or "Sequoia networking issues" will show plenty of issues. 

Not saying that this has anything to do with what the OP is running into, but it bears at least some consideration. Definitely impacting NAS usage as well. 

The more interesting issue I had (still have!) is iMessage not working with VPN on one Mac, and only this one Mac. Same OS and VPN (and version) on all other Macs, no issue, but for this Mac, iMessage just doesn't work. For a while it worked when I turned the VPN off, but then it stopped working at that point too. I escalated to Apple at that point and my call was escalated within Apple, it was a 2+ hour data collection fun-fest. They know and acknowledge this is a bug, but as of 2 weeks ago, they were still very confused as to what was happening. 

Folks are pointing to the firewall, but during the call we turned it on and off and lots of other combinations and it had no effect. 

While yes this is an Internet issue, I've run into odd LAN issues as well that are hard to quantify. 15.0.1 did not fix any of the issues I have or have seen.

**Reply by RhetTbull** [reply]
@cfc62 that's why I'm not on Sequoia yet and why OSXPhotos isn't fully tested on Sequoia. I prefer to stay one version behind.

**Comment by gerdemb** [comment]
Here's the end of the log of my last run with the profiling data. I am not using Sequoia. The `--only-new` flag does sound closer to what I want. I will try another run after adding it.

> Exports are resumable. But OSXPhotos still needs to check each photo to see if something has changed between Photos and the exported files. For example, you edited a photo or added a keyword. My priority is maximum preservation of data, not speed. To do this OSXPhotos must look at the exported file to see if it's different than expected. Over a slow network/NAS this takes time. OSXPhotos is not optimized for this use case.

I can see how this would hurt performance especially when connecting to a NAS. If you made the assumption that OSXPhotos was the only writer to the export directory and no modifications were being made to files by the user you could you rely entirely on the exportedb without checking the photo files to see if it has changed?

> As for never finishing, I don't know what the issue is but if there is a bug I'd definitely like to squash it. Unfortunately without being able to replicate it it's hard to know what the issue is. If OSXPhotos crashes it creates a crash log with debug info but it sounds like in your case it's just hanging.

To be clear I didn't meant that the backup hung or crashed, rather that I was never able to run a en export to completion after multiple runs over many days. Due to the way progress is reported it was difficult to tell how much of the library has already been exported and it _seems_ that that on a NAS resuming an export is not significantly faster than the first export. It sounds like `--only-new` flag might resolve this issue though.

```
osxphotos \
  --profile \
  export \
  /Volumes/photo/osxphotos \
    --export-by-date \
  --sidecar xmp \
  --update \
  --export-aae \
  --person-keyword \
  --download-missing \
  --use-photokit \
  --ramdb \
  --exportdb ~/osxphotos_export.db \
  --ignore-signature \
  --checkpoint 100 \
  --timestamp \
  --limit 5000
  ```
  
```
2024-10-25 14:43:46.639189 -- Processed: 267296 photos, exported: 5253, updated: 19, skipped: 309, updated EXIF data: 0, missing: 8, error: 0, limit: 5000/5000 exported
2024-10-25 14:43:46.640311 -- Elapsed time: 3:57:04
2024-10-25 14:43:46.707821 -- Writing export database changes back to /Users/bengerdemann/osxphotos_export.db
2024-10-25 14:43:47.472596 -- Cleaning up lock files
,Profiling completed
         535655871 function calls (523760989 primitive calls) in 17784.213 seconds

```

[profile.txt](https://github.com/user-attachments/files/17529683/profile.txt)

**Reply by oPromessa** [reply]
It seems it ran flawlessly for the 5k limit. 

> To be clear I didn't meant that the backup hung or crashed, rather that I was never able to run a en export to completion after multiple runs over many days. Due to the way progress is reported it was difficult to tell how much of the library has already been exported and it seems that that on a NAS resuming an export is not significantly faster than the first export. It sounds like --only-new flag might resolve this issue though.

As for progress info:
- If you run osxphotos export from the terminal in an interactive session it will show you a progress bar at the bottom. Unless you use the --no-progress option. 
- if you redirect output to a file say osxphotos.log, you can see progress by running command "tail -f osxphotos.log" and you can see indication of progress "(55321/260765)". You can press Ctrl-C to cancel "tail" command. 
- just let it run... and check progress with tail. Try --limit 100000 or even remove it. Have faith ;)

**Reply by RhetTbull** [reply]
> I can see how this would hurt performance especially when connecting to a NAS. If you made the assumption that OSXPhotos was the only writer to the export directory and no modifications were being made to files by the user you could you rely entirely on the exportedb without checking the photo files to see if it has changed?

Yes, if you assumed no other process/person was writing to the directory and that you started with an empty directory, this could be optimized. Today, OSXPhotos lists the contents of the export directory to see if there's a possible name collision when writing a new file. You would still need to check for name collisions but the code could do this by querying the export database to see if a photo with that name was already exported and this would be faster than a directory read over a NAS. I'll open an issue to add an option for this but not sure when I'll get to it.

**Reply by gerdemb** [reply]
> Today, OSXPhotos lists the contents of the export directory to see if there's a possible name collision when writing a new file. You would still need to check for name collisions but the code could do this by querying the export database to see if a photo with that name was already exported and this would be faster than a directory read over a NAS. I'll open an issue to add an option for this but not sure when I'll get to it.

Doesn't the `--update-new` flag already cover this case? Adding it improved the performance of my export and also reported progress incrementally showing the number of remaining items subtracting those that had already been exported.

For future reference and anyone else following along, here are the final command-line switches I settled on:

```
osxphotos \
  --profile \
  export \
  /Volumes/photo/osxphotos \
  --export-by-date \
  --sidecar xmp \
  --update \
  --export-aae \
  --person-keyword \
  --download-missing \
  --use-photokit \
  --ramdb \
  --exportdb ~/osxphotos_export.db \
  --ignore-signature \
  --timestamp \
  --limit 5000 \
  --only-new \
  --verbose --verbose
```

**Reply by RhetTbull** [reply]
`--update --only-new` still checks for name collisions by listing the directory contents but it's only doing it for photos not yet exported. As you noted this will be much faster as the exported percentage grows. What I was thinking is that if there was an option that used the export database as source of truth instead of the actual directory contents, then the code could do a fast SQL query to see if there's a name collision and the only operation over the network to the NAS would be the actual write.

**Reply by oPromessa** [reply]
That would be great. 

Just wondering the delete operation to "cleanup files previously exported which were in the meanwhile deleted from Photos" would continue to work/applied, correct?

**Reply by RhetTbull** [reply]
`--cleanup` does not currently work as you might expect with `--only-new`. It will delete an files previously exported in this case. I'll open a new issue for this. Once fixed, then yes, I'd expect `--cleanup` to work when using the new `--blind-copy` option.

**Reply by RhetTbull** [reply]
See #1727 and #1505

**Comment by rajscode** [comment]
@gerdemb Not sure if this will be helpful at all as it seems you already have optimized speed / connection between your NAS and Mac.  If you haven't you may want to see some of the info in these two posts:  [Post 1](https://github.com/RhetTbull/osxphotos/discussions/1415#discussioncomment-8868921) and [Post 2](https://github.com/RhetTbull/osxphotos/discussions/1415#discussioncomment-9000122) especially if you are using SMBv1.

**Reply by gerdemb** [reply]
@rajscode thanks for the suggestions. I am sure that are some things I could do to try and speed up the data transfer between my Mac and NAS, but I run some other jobs with large data transfers and none of them run nearly as slowly as `osxphotos`. In other words, I don't think bottleneck is the NAS, but rather the processing that `osxphotos` is doing on the Mac itself. It's also possible that `osxphotos` has some disk access patterns that are performant when working on a local drive, but not when accessing data over a network. For example, the initial directory walk of the destination directory before the export starts seems to be very slow (ten minutes or more maybe?), but I imagine the same access on a local disk would be an order of magnitude or so faster.

**Reply by rajscode** [reply]
Efficient vs. Large Transfers. Efficient file transfer with osxphotos isn’t just about total data size. For example, transferring 500 files totaling 2TB may work smoothly, but transferring 200,000+ smaller files can become inefficient due to configuration issues like:

- Using SMB1 instead of SMB2/3
- Limited client-side cache size (macOS) for directory and file operations (i.e. SMB1)
- Insufficient cache on the NAS for directory walking (i.e. SMB1). Details of this are in posts I refered to originally.

Even with a local SSD storing an optimized iCloud Photos library, exports via osxphotos are inherently slow because each run downloads every photo from iCloud.com.

My Setup for a Large Library (160K+ Photos, 6.5K Videos, 1.7TB)

To maintain a full-resolution backup outside the Photos app:

1. Primary Login (MacBook Pro, 1TB SSD): Optimized library stored internally.
2. Secondary Login (Same Mac): Full-resolution Photos library stored on a 2TB USB-C SSD.
3. Weekly osxphotos Export: Runs from the secondary login to the same external SSD. Thanks to APFS, exported files (~1.7TB) efficiently coexist with the Photos library.

This structure lets me manage my media in a standard directory format, enabling easy backups to a NAS or cloud storage.

For details, see posts on MacRumors: ([High Level Backup Stragegy](https://forums.macrumors.com/threads/backup-strategy-optimized-icloud-photo-library-on-mac-and-full-one-on-external-ssd.2351777/?post=31738760#post-31738760) and [Details of Dual Library on a single Mac](https://forums.macrumors.com/threads/backup-solution-icloud-full-photo-library-on-external-drive-whilst-optimized-reduced-version-on-internal-drive.2369845/))

EDIT ADD:  
- To be clear I've everything up in iCloud and the storage used there for Photos App is 1.7TB. 
- osxphotos export: as a weekly scheduled job takes approx 1hr30min to complete for me on a weekly basis now.

**Reply by RhetTbull** [reply]
> In other words, I don't think bottleneck is the NAS, but rather the processing that osxphotos is doing on the Mac itself. It's also possible that osxphotos has some disk access patterns that are performant when working on a local drive, but not when accessing data over a network.

This is true. OSXPhotos will stage each file to a temporary directory before it processes it. This vastly simplifies the internal logic and if on a local APFS disk, there's no performance penalty due to copy-on-write. If Photo library is on a different disk or across the network, this slows things down a lot. If the file has not been downloaded from iCloud it will also have to be downloaded then staged which further slows things down. OSXPhotos is not yet optimized to work with "optimized" libraries though I'm slowly making progress to improve this. See #1086.

Secondly, OSXPhotos does a "stat" on the target directory for every single file that is copied. Over a network this can be very slow, especially if there are lots of files in the target directory. The reason for this is to ensure the desired target name is available (doesn't already exist on the target volume) and to see if the target copy differs from expected and needs to be updated (this part can be skipped with `--ignore-signature`).  I've opened #1723 recently to add a `--blind-copy` mode that will assume the targetfile name is available (overwriting it if present) which should speed up network export significantly. The caveat will be that you cannot modify files in the export directory without risk of corruption. With this mode, it must be assumed that OSXPhotos is the only process that will ever modify the target directory.

**Reply by cfc62** [reply]
@gerdemb Completely outside of Photos/Osxphotos I do keep track of other assets and _absolutely_ there can be (and are) very very large disparities in performance networked vs. direct attached storage. 

As an example if I open up this other database with the assets available over the network, time from start to a working session can be 4-5 minutes and more, where if I open up the database with the assets directly attached, it takes 15-20 seconds at most. That's perhaps 20 to 30x faster on direct attached storage.

It's not related network load in any way, neither the network accesses nor the disk accesses are anything substantial at all. 

PS: the network is fully wired, no wi-fi inconsistencies involved. It's 1G, based on what I've seen I doubt going to either 2.5G or 10G would make much of a difference.
